### PR TITLE
organize test log

### PR DESCRIPTION
**Pull request by jenniexie**
organize test logs to under build folder by using “abs_logdir” in os.environ, there are 3 category logs:
1. Device(happy node) logs handled in happy/HappyProcessStart.py
2. Happy_debug_log.txt handled in Driver.py 
3. Persistent log handled in WeaveTest.py (this will be addressed in another openweave PR)

The abs_logdir value will be assigned from setup() in test script, and will be covered in another PR in openweave-core.
example code in openweave-core to assign value to abs_logdir in test script:

 self.feature_dir = re.findall(
            r'.*openweave-core/src/test-apps/happy/tests/(.*)', os.path.dirname(os.path.abspath(__file__)))[0]

os.environ['abs_logdir'] = os.environ['abs_builddir'] + '/happy/test_logs/' + self.feature_dir + '/'

All changes have been tested and verified OK.
PR from openweave-core related to this feature:
https://github.com/openweave/openweave-core/pull/265

### CODE REVIEW

**Comment by jenniexie** [comment]
Based on discussion with @robszewczyk today, I have made below changes:
1. introduced HAPPY_LOG_DIR in weave-parallel-test-engine as os environment variable
2.use HAPPY_LOG_DIR for both happy device log and happy host log in Driver.py
3. changed main_conf and log_conf to use happy_log_dir

There will be another openweave-core PR needed to use the feature of HAPPY_LOG_DIR

**Review by gerickson — CHANGES_REQUESTED** [review]
It feels like there should be a high-level documentation change that precedes this check in that describes, in plain, easy-to-understand terms how Happy logging works and then the code should be cross-checked against that description.

The documentation should also pass the @Vyrastas test.

**Comment by jenniexie** [comment]
> It feels like there should be a high-level documentation change that precedes this check in that describes, in plain, easy-to-understand terms how Happy logging works and then the code should be cross-checked against that description.
> 
> The documentation should also pass the @Vyrastas test.

Thanks Grant for your feedback!
Based on our discussion, I will add doc/Logging.md to happy repo to explain how the logging works inside Happy, and will add Jeff to my reviewer for the doc.

**Comment by Vyrastas** [comment]
Just let me know when it's ready and I'll review, and yes please put it in a new /doc folder.

**Comment by jenniexie** [comment]
> Just let me know when it's ready and I'll review, and yes please put it in a new /doc folder.

Thanks Jeff@Vyrastas , I have uploaded Logging.md under doc/, could you help to review and let me know, Thanks!

File: doc/Logging.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+How does Happy test log work
```
Comment (Vyrastas) [inline doc/Logging.md]: Let's make the filename lowercase logging.md

I also have a general documentation style guide under review that you might want to reference, let's start keeping everything the same style as much as possible: https://github.com/openweave/openweave-core/blob/84721a01b2a4c0b1476e0059b7f185b19c4c7955/doc/STYLE_GUIDE.md

And use a title header and remove line 2

# How Happy test logs work

File: doc/Logging.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+============================
+
+## Happy produced logs
+  * Happy host logs
```
Comment (Vyrastas) [inline doc/Logging.md]: Let's describe what types of logs are produced, specifically.  What's the difference between host and device?

File: doc/Logging.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  * Happy device logs
+
+## Where to store Happy logs
+
+Both Happy host and device logs location are processed in happy/Driver.py as below:
+  * if HAPPY_LOG_DIR is found in os environment variable, happy uses HAPPY_LOG_DIR's value as log folder location to store happy test logs. For example, openweave-core tests store happy test logs along with feature folder under build folder.
```
Comment (Vyrastas) [inline doc/Logging.md]: - Let's explain how to set HAPPY_LOG_DIR as an OS environment variable.
- Render all the code terms in code font.  So like `HAPPY_LOG_DIR`.
- Call it "Happy" (capitalized, no code font) through out the doc unless you're referring to the command itself
- "...Happy uses `HAPPY_LOG_DIR` as the log folder location..."
- Capitalize each bullet and end with a period.

File: doc/Logging.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  * if HAPPY_LOG_DIR is not found in os environment variable, happy uses "default_happy_log_dir" from happy/conf/main_config.json as default log folder location to store happy test logs
```
Comment (Vyrastas) [inline doc/Logging.md]: - `default_happy_log_dir` in code font
- Link to happy/conf/main_config.json in the repo
- ...as **the** default log folder..."
- Capitalize bullet and end with a period.

File: doc/Logging.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+How does Happy test log work
+============================
+
+## Happy produced logs
```
Comment (Vyrastas) [inline doc/Logging.md]: Let's change this to

## Types of Happy logs

File: doc/Logging.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  * Happy host logs
+  * Happy device logs
+
+## Where to store Happy logs
```
Comment (Vyrastas) [inline doc/Logging.md]: ## Where Happy logs are stored

File: doc/Logging.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+Both Happy host and device logs location are processed in happy/Driver.py as below:
+  * if HAPPY_LOG_DIR is found in os environment variable, happy uses HAPPY_LOG_DIR's value as log folder location to store happy test logs. For example, openweave-core tests store happy test logs along with feature folder under build folder.
+  * if HAPPY_LOG_DIR is not found in os environment variable, happy uses "default_happy_log_dir" from happy/conf/main_config.json as default log folder location to store happy test logs
+
+Happy host logs location variable are used by python logging module to store happy host logs.
```
Comment (Vyrastas) [inline doc/Logging.md]: For these two lines (14 & 15), link to the place in the repo where each module is storing the logs.  We could probably just have this as a list:

* Happy host logs are processed by <python logging module>
* Happy device logs are processed by <happy/HappyProcessStart.py>

File: doc/Logging.md (hunk 4 above)
Comment (Vyrastas) [inline doc/Logging.md]: Capitalize bullet and end with a period.

File: doc/Logging.md (hunk 1 above)
Comment (jenniexie) [reply doc/Logging.md]: done

File: doc/Logging.md (hunk 2 above)
Comment (jenniexie) [reply doc/Logging.md]: done

File: doc/Logging.md (hunk 3 above)
Comment (jenniexie) [reply doc/Logging.md]: done

File: doc/Logging.md (hunk 4 above)
Comment (jenniexie) [reply doc/Logging.md]: done

**Comment by jenniexie** [comment]
Hi Jeff @Vyrastas:
    Thanks for your feedback! I have addressed all comments, could you check again and let me know?

**Review by Vyrastas — APPROVED** [review]
Thanks for the updates! This is a good starting point. There are still some minor wording and style things we should fix... but @jenniexie and I talked offline and decided it would be easier to merge as is, and then I'll do a quick PR to fix the items and align the doc with the new doc style we're implementing for OpenWeave repos.  If other approvers are okay with this, please merge and I'll follow up with a PR.

**Comment by gerickson** [comment]
> Thanks for the updates! This is a good starting point. There are still some minor wording and style things we should fix... but @jenniexie and I talked offline and decided it would be easier to merge as is, and then I'll do a quick PR to fix the items and align the doc with the new doc style we're implementing for OpenWeave repos. If other approvers are okay with this, please merge and I'll follow up with a PR.

I'll take a look at the technical details this afternoon.

File: doc/logging.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# How does Happy test log work
+
+## Types of Happy logs
+  * Happy host logs — Happy will run as host, Happy host logs will have logs from Happy tool only.
+
+  * Happy device logs — Happy will host virtual network which has multiple virtual devices, Happy device logs will have logs for software running on the virtual device, openweave is one of the software example, so device logs will have openweave logs.
```
Comment (gerickson) [inline doc/logging.md]: When we mention "openweave" we should be consistent in our syntax, for example "openweave" vs. "openweave-core".

File: doc/logging.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# How does Happy test log work
+
+## Types of Happy logs
+  * Happy host logs — Happy will run as host, Happy host logs will have logs from Happy tool only.
```
Comment (gerickson) [inline doc/logging.md]: What do we mean by "host" here or is that introduced in other documentation? If not, we should clarify that Happy's role as "host" is that of test coordination and orchestrator, facilitating the activities of the virtual test device nodes.

File: doc/logging.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+  * Happy device logs — Happy will host virtual network which has multiple virtual devices, Happy device logs will have logs for software running on the virtual device, openweave is one of the software example, so device logs will have openweave logs.
+
+## Where Happy logs are stored
+
+Both Happy host and device logs location are processed in `happy/Driver.py` as below:
+  * If `HAPPY_LOG_DIR` is found in os environment variable, Happy uses `HAPPY_LOG_DIR` value as log folder location to store Happy test logs. For example, openweave-core tests store Happy test logs along with feature folder under build folder.
```
Comment (gerickson) [inline doc/logging.md]: When we mention "openweave" we should be consistent in our syntax, for example "openweave" vs. "openweave-core".

File: doc/logging.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  * If `HAPPY_LOG_DIR` is not found in os environment variable, Happy uses `default_happy_log_dir` from [happy/conf/main_config.json](https://github.com/openweave/happy/blob/master/happy/conf/main_config.json) as the default log folder location to store Happy test logs.
```
Comment (gerickson) [inline doc/logging.md]: Is this something the user can or should control or modify? What is an example of the default value associated with the key 'default_happy_log_dir'? Is it /tmp/<something>? Is it ${HOME}/<something>?

File: doc/logging.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+## How to set `HAPPY_LOG_DIR`
+  There is multiple ways to set `HAPPY_LOG_DIR` in os environment variable, see examples below:
+  * In openweave test, we add `HAPPY_LOG_DIR=<dir value>` in command line along with the test script we want to run in [weave-parallel-test-engine](https://github.com/openweave/openweave-core/blob/master/src/test-apps/weave-parallel-test-engine)
```
Comment (gerickson) [inline doc/logging.md]: This doesn't belong here. The Happy package shouldn't be aware of how packages that depend on it are using it. Consequently, it shouldn't link to the openweave-core project either.

A better and less-coupled example might be: "If your project is being built with GNU autotools and you are running functional and unit tests with Happy with the DejaGnu test framework, then you might add HAPPY_LOG_DIR=<dir value> to TESTS_ENVIRONMENT in your Makefile.am or Makefile.in.

File: doc/logging.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+  * We can add `HAPPY_LOG_DIR=<dir value>` in .bashrc, and source it before we run the test script.
+
+## Where is logs location variable being used
+* Happy host logs location variable are used by python logging module to store Happy host logs in [Driver.py](https://github.com/openweave/happy/blob/master/happy/Driver.py#L223)
```
Comment (gerickson) [inline doc/logging.md]: Do users of Happy need to know this or do developers and maintainers of Happy need to know this?

The documentation seems to co-mingle information for users and developers. However, the documentation should keep those roles and concerns separate. The user wants to know "How do I just use Happy and get my job done with it?". The developer wants to know, "How do I maintain and extend Happy  to accomplish a new task or implement a new feature?".

**Review by gerickson — CHANGES_REQUESTED** [review]

File: doc/logging.md (hunk 11 above)
Comment (jenniexie) [reply doc/logging.md]: Right, user can modify default_happy_log_dir if they want.
'default_happy_log_dir' default value is /tmp in main_config.json
added information in the PR

File: doc/logging.md (hunk 13 above)
Comment (jenniexie) [reply doc/logging.md]: Users don't need to know this part information, how about I delete this part. We can think about do we need to keep the information somewhere and where to keep.

File: doc/logging.md (hunk 12 above)
Comment (jenniexie) [reply doc/logging.md]: sure, I will remove how openweave-core tests using this variable.

File: doc/logging.md (hunk 8 above)
Comment (Vyrastas) [reply doc/logging.md]: This is tricky because if I were referring to the repo itself I'd just use openweave-core... but if we're talking about the high-level "product" is there a difference between OpenWeave and OpenWeave Core?  If there is no meaningful difference I think we should just use OpenWeave for the product in general and openweave-core when referring to the repo specifically in code examples.

File: doc/logging.md (hunk 9 above)
Comment (jenniexie) [reply doc/logging.md]: Changed "Happy host logs" to "Happy debug logs", explained Happy's role in the test too.

File: doc/logging.md (hunk 10 above)
Comment (jenniexie) [reply doc/logging.md]: changed to openweave in all places

File: doc/logging.md (hunk 8 above)
Comment (jenniexie) [reply doc/logging.md]: How about we use OpenWeave in all places

**Comment by jenniexie** [comment]
Thanks Grant @gerickson and Jeff @Vyrastas for your feedback, I have addressed all comments now.
Please help to take a look one more time, Thanks!

File: doc/logging.md [hunk 14]
```diff
@@ -0,0 +1,17 @@
+# How does Happy test log work
+
+## Types of Happy logs
+  * Happy debug logs — Happy will run as test coordination and orchestrator, facilitating the activities of the virtual test device nodes. Happy debug logs will have logs from Happy tool only.
+
+  * Happy device logs — Happy will host virtual network which has multiple virtual devices, Happy device logs will have logs for software running on the virtual device, OpenWeave is one of the software example, so device logs will have OpenWeave logs.
```
Comment (gerickson) [inline doc/logging.md]: will have -> may have?

File: doc/logging.md [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+## Where Happy logs are stored
+
+Both Happy host and device logs location are processed in `happy/Driver.py` as below:
```
Comment (gerickson) [inline doc/logging.md]: "...are processed in `happy/Driver.py` as below" seems like an implementation detail that should be elided here.

File: doc/logging.md [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+  * If `HAPPY_LOG_DIR` is found in os environment variable, Happy uses `HAPPY_LOG_DIR` value as log folder location to store Happy test logs. For example, OpenWeave tests store Happy test logs along with feature folder under build folder.
```
Comment (gerickson) [inline doc/logging.md]: What is a "feature folder"?

File: doc/logging.md [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  * If `HAPPY_LOG_DIR` is not found in os environment variable, Happy uses `default_happy_log_dir` value from [happy/conf/main_config.json](https://github.com/openweave/happy/blob/master/happy/conf/main_config.json) as the default log folder location to store Happy test logs. The default value of `default_happy_log_dir` is `/tmp`, user can modify it if they want.
```
Comment (gerickson) [inline doc/logging.md]: Thank you for clarifying the default value of default_happy_log_dir in main_config.json. However, if someone is using an installed Happy distribution, main_config.json isn't something they can easily override, correct? This is really only something someone could / should modify if they were working on a "source" distribution of Happy, correct?

If so, this again seems like an implementation detail that shouldn't be included in this section of the documentation.

File: happy/conf/main_config.json [hunk 18]
```diff
@@ -1,10 +1,12 @@
 {
     "configuration_file":"~/.happy_conf.json",
     "log_directory":"/tmp",
+    "happy_log_environ":"HAPPY_LOG_DIR",
```
Comment (gerickson) [inline happy/conf/main_config.json]: Is there a strong value in adding this additional level of indirection to determine the name of the environment variable to read for the Happy log directory?

File: happy/Driver.py [hunk 19]
```diff
@@ -195,6 +195,28 @@ def __configure(self):
...
+        # default_happy_log_dir will be default dir for happy log
+        # happy_log_environ is os environment variable passed from test-engine
+        try:
+            self.default_happy_log_dir = self.main_conf["default_happy_log_dir"]
+        except Exception:
+            print "Failed to find default_happy_log dir in the main configuration file %s." \
+                % (self.main_conf_file)
+            sys.exit(1)
+
+        try:
```
Comment (gerickson) [inline happy/Driver.py:207]: Is there a strong value in adding this additional level of indirection to determine the name of the environment variable to read for the Happy log directory?

File: doc/logging.md [hunk 20]
```diff
@@ -0,0 +1,17 @@
+# How does Happy test log work
+
+## Types of Happy logs
+  * Happy debug logs — Happy will run as test coordination and orchestrator, facilitating the activities of the virtual test device nodes. Happy debug logs will have logs from Happy tool only.
```
Comment (Vyrastas) [inline doc/logging.md]: Let's reword this a little so we're not repeating "Happy" so much

* Debug logs - Logs related to Happy's role as test coordinator and orchestrator, where it facilitates the activities of virtual device nodes.

File: doc/logging.md (hunk 14 above)
Comment (Vyrastas) [inline doc/logging.md]: We can reword this too.  As for "software" does that mean anything that can be running on the node?  Any Linux binary visible to the namespace, or just stuff related to Happy commands?

* Device logs - Logs related to software running on individual virtual device nodes within a Happy test network.

File: happy/Driver.py (hunk 19 above)
Comment (jenniexie) [reply happy/Driver.py:207]: Yup, by adding this environment variable, we can organize openweave/happy tests logs, we can specify where to save every test device log and happy debug log. Before, everything is under /tmp, no organization for the device logs and happy debug log.

FYI: here is how the environment variable used in openweave test launcher:
https://github.com/openweave/openweave-core/pull/265/files#diff-7dc13b60eed8ccc1a74c2d26886ed8e7

HAPPY_LOG_DIR is similar implementation as HAPPY_STATE_ID for Happy

File: doc/logging.md (hunk 20 above)
Comment (jenniexie) [reply doc/logging.md]: OKay

File: doc/logging.md (hunk 14 above)
Comment (jenniexie) [reply doc/logging.md]: ok

File: happy/Driver.py (hunk 19 above)
Comment (gerickson) [reply happy/Driver.py:207]: To be clear, I'm not advocated not having HAPPY_LOG_DIR. The question I am raising is whether there is any value in having 'HAPPY_LOG_DIR' as the name of the environment variable to be used to be configurable and loadable via `self.happy_log_environ = self.main_conf["happy_log_environ"]`? I don't think there is and believe the code can and should be just:

```
self.happy_log_environ = "HAPPY_LOG_DIR"
```

File: doc/logging.md (hunk 14 above)
Comment (jenniexie) [reply doc/logging.md]: Hi Jeff: 
    The "software" here means "openweave", how about change to "software such as openweave"

File: happy/Driver.py (hunk 19 above)
Comment (jenniexie) [reply happy/Driver.py:207]: > To be clear, I'm not advocated not having HAPPY_LOG_DIR. The question I am raising is whether there is any value in having 'HAPPY_LOG_DIR' as the name of the environment variable to be used to be configurable and loadable via `self.happy_log_environ = self.main_conf["happy_log_environ"]`? I don't think there is and believe the code can and should be just:
> 
> ```
> self.happy_log_environ = "HAPPY_LOG_DIR"
> ```

Thanks Grant for your feedback!
I have removed the option from the config file.

File: doc/logging.md (hunk 14 above)
Comment (jenniexie) [reply doc/logging.md]: done

File: doc/logging.md (hunk 15 above)
Comment (jenniexie) [reply doc/logging.md]: removed that line

File: doc/logging.md (hunk 20 above)
Comment (jenniexie) [reply doc/logging.md]: done

**Comment by jenniexie** [comment]
I have addressed all comments, please let me know if you have other comments, otherwise I need green lights from you :)

File: doc/logging.md [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+
+## Types of Happy logs
+
+  * Debug logs - Logs related to Happy's role as test coordinator and orchestrator, where it facilitates the activities of virtual device nodes.
+
+  * Device logs - Logs related to software such as openweave running on individual virtual device nodes within a Happy test network. OpenWeave is one of the software example, device logs will have OpenWeave logs.
+
+## Where Happy logs are stored
+
+  * If `HAPPY_LOG_DIR` is found in os environment variable, Happy uses `HAPPY_LOG_DIR` value as log folder location to store Happy test logs.
```
Comment (Vyrastas) [inline doc/logging.md]: If the `HAPPY_LOG_DIR` environment variable exists, Happy uses the `HAPPY_LOG_DIR` value...

File: doc/logging.md [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+
+  * If `HAPPY_LOG_DIR` is not found in os environment variable, Happy uses `default_happy_log_dir` value from [happy/conf/main_config.json](https://github.com/openweave/happy/blob/master/happy/conf/main_config.json) as the default log folder location to store Happy test logs. The default value of `default_happy_log_dir` is `/tmp`.
```
Comment (Vyrastas) [inline doc/logging.md]: If the `HAPPY_LOG_DIR` environment variable doesn't exist, Happy uses the `default_happy_log_dir` value...

File: doc/logging.md [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+
+## How to set `HAPPY_LOG_DIR`
+
+  There is multiple ways to set `HAPPY_LOG_DIR` in os environment variable, see examples below:
```
Comment (Vyrastas) [inline doc/logging.md]: There are multiple ways to see the `HAPPY_LOG_DIR` environment variable:

(remove "see examples below")

File: doc/logging.md [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+  * If your project is being built with GNU autotools and you are running functional and unit tests with Happy with the DejaGnu test framework, then you might add `HAPPY_LOG_DIR=<dir value>` to TESTS_ENVIRONMENT in your Makefile.am or Makefile.in.
+  * We can add `HAPPY_LOG_DIR=<dir value>` in .bashrc, and source it before we run the test script.
```
Comment (Vyrastas) [inline doc/logging.md]: Add `HAPPY_LOG_DIR=<dir value>` to .bashrc and `source` it before you run the test script.

**Review by Vyrastas — CHANGES_REQUESTED** [review]
Just a few more slight rewordings.

**Comment by jenniexie** [comment]
@Vyrastas Hi Jeff, I addressed your comments, please check it again :)

**Review by Vyrastas — APPROVED** [review]

File: doc/logging.md [hunk 25]
```diff
@@ -0,0 +1,19 @@
+# How does Happy test log work
+
+## Types of Happy logs
+
+  * Debug logs - Logs related to Happy's role as test coordinator and orchestrator, where it facilitates the activities of virtual device nodes.
+
+  * Device logs - Logs related to software such as openweave running on individual virtual device nodes within a Happy test network. OpenWeave is one of the software example, device logs may have OpenWeave logs.
```
Comment (gerickson) [inline doc/logging.md]: We can clean it up in a subsequent pass; however, the capitalization of OpenWeave here is inconsistent.

Comment (jenniexie) [reply doc/logging.md]: good eyes:)
I fixed it to "OpenWeave"

**Review by gerickson — APPROVED** [review]

**Review by robszewczyk — APPROVED** [review]
